zerver/tests/test_upload.py

```diff
@@ -142,6 +181,38 @@ def test_multiple_claim_attachments(self):
 
         self.assertEquals(Attachment.objects.get(path_id=d1_path_id).messages.count(), 2)
 
+    def test_file_download_authorization(self):
+        subscribed_users = ["hamlet@zulip.com", "iago@zulip.com"]
+        unsubscribed_users = ["othello@zulip.com", "prospero@zulip.com"]
+        for user in subscribed_users:
+            self.subscribe_to_stream(user, "test-subscribe")
+
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client.post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
+        fp_path_id = re.sub('/user_uploads/', '', uri)
+        body = "First message ...[zulip.txt](http://localhost:9991/user_uploads/" + fp_path_id + ")"
+        self.send_message("hamlet@zulip.com", "test-subscribe", Recipient.STREAM, body, "test")
+        self.client.post('/accounts/logout/')
+
+        # Subscribed user should be able to view file
+        for user in subscribed_users:
+            self.login(user)
+            response = self.client.get(uri)
+            data = "".join(response.streaming_content)
+            self.assertEquals("zulip!", data)
+            self.client.post('/accounts/logout/')
+
+        # Unsubscribed user should not be able to view file
+        for user in unsubscribed_users:
+            self.login(user)
+            response = self.client.get(uri)
+            self.assert_json_error(response, "You are not authorized to view this file.", status_code=403)
+            self.client.post('/accounts/logout/')
+
```

Comment:
Can you add a test for the 404 case in here somewhere?  I think the recipe of uploading a file, `rm`ing it on disk, and then trying to view it would work.